### ISSUE TITLE
sset for JSON column [Postgres]

### ISSUE THREAD

**Issue by RyanBertrand** [jkk/honeysql#155]
Hey everyone,

I am trying to update a column which is of type json. 

```clojure
(defn- update-where
"Builds a HoneySQL map to update-where"
[table row-data where]
(-> (update table)
      (sset row-data)
      (where where)))
```

```clojure
(-> (update-where :my_table {:json_col_name {:fake "data"}} [:= :id 1])
      (build)
      (format :quoting :ansi)))
```

Returns:
["UPDATE \"my_table\" SET \"json_col_name\" = () WHERE \"id\" = ?" 1]

It seems that the JSON (hash-map) just gets wiped out.  Is this a bug or works as expected? 

Thanks,
Ryan

**Comment by jrdoane** [comment]
HoneySQL has no way of knowing which database the query is going to be executed against as most options relate to how certain things are done, like quoting and such. I've gone down this road before and I think it's safe to say that for things like this, using clojure.java.jdbc directly. I've successfully does this for the json field and for a custom "edn" data type that uses nippy to serialize and deserialize data and use it directly with PostgreSQL using clojure.java.jdbc directly.

tl;dr: This is probably outside of the scope of HoneySQL as a library.

**Comment by RyanBertrand** [comment]
Ah okay makes sense - thanks for your response.

**Comment by molsder** [comment]
I got around this by setting the value using raw sql:

`(defn- jset [m field value]
  (if (some? value)
    (assoc-in m [:set field] (sql/raw "?::json"))
    m))`

After the sset call, I call jset for the json column, then providing a JSON string representation as a parameter.

**Comment by emidln** [comment]
The value is being interpreted as a sub-query since it's a map. To avoid this, wrap the dict `honeysql.format/value` and extend `clojure.java.jdbc/ISQLValue` for `clojure.lang.IPersistentMap`  ala:

```clj
(extend-protocol clojure.java.jdbc/ISQLValue
  clojure.lang.IPersistentMap
  (sql-value [value] (doto (PGObject.) (.setType "jsonb") (.setValue (cheshire.core/generate-string value))))
```

**Comment by manuel-uberti** [comment]
Thanks @emidln, you saved my day :)

**Comment by danielcompton** [comment]
I went a different way around this by casting the JSON string to JSONB with `honeysql.core/call`:

```
(jdbc/execute!
  tx
  (-> (sql/insert-into :my_table)
      (sql/values [{:json_payload (h/call :cast json-string :jsonb)}])
      h/format))
```

This doesn't give you the easy coercion between Clojure maps to JSON datatypes but does give you more precision over casts and would prevent accidental casts. You'd probably want something else on the read side for reading back out.
